metric_learn/lmnn.py

```diff
@@ -158,14 +162,25 @@ def fit(self, X, y):
       if delta_obj > 0:
         # we're getting worse... roll back!
         learn_rate /= 2.0
+        L = L_old
+        G = G_old
         df = df_old
         a1 = a1_old
         a2 = a2_old
         objective = objective_old
       else:
-        # update L
-        L -= learn_rate * 2 * L.dot(G)
-        learn_rate *= 1.01
+        # We did good. We store this point as reference in case we do
+        # worse next time.
+        objective_old = objective
+        L_old = L.copy()
+        G_old = G.copy()
+        df_old = df.copy()
+        a1_old = [a.copy() for a in a1]
+        a2_old = [a.copy() for a in a2]
+      # we update L and will see in the next iteration if it does indeed
+      # better
+      L -= learn_rate * 2 * L.dot(G)
+      learn_rate *= 1.01
```

Comment:
This seems wrong to me, though it was also wrong before this PR.

Reading through the Shogun implementation [here](http://shogun-toolbox.org/api/latest/LMNN_8cpp_source.html) and [here](http://shogun-toolbox.org/api/latest/LMNNImpl_8cpp_source.html), they don't do any rolling back.
They do the `L` update unconditionally in `gradient_step`, then compute the objective value for the current iteration, then do the learning rate update based on the change in objective.

In one of the reference Matlab implementations [here](https://bitbucket.org/mlcircus/lmnn/src/5b49cafaeb9aacf19571f0b334178830cf712056/lmnn2/lmnn2.m?at=default&fileviewer=file-view-default#lmnn2.m-234) they do the `L` update first, then optionally roll back to a saved state when updating the step size.

--
So I think the correct fix would be to move the `L` update to the `# do the gradient update` section, after computing the new `G`, using the existing learning rate. Then, if the objective didn't improve we can halve the learning rate and roll back to the last good state (including `L` and `G`). Otherwise, we just grow the learning rate by 1% and carry on.

Reply:
Inverting the order of things as suggested would also improve the readability of the code I think

Reply:
I just submitted a new commit inverting things, I'll comment the lines to explain the change